### ISSUE TITLE
Options for integration tests

### ISSUE THREAD

**Issue by alvipeo** [aspnet/Docs#8130]
It's not clear to me how can I use Options pattern for tests. Let's say I run tests on my dev machine and then I have CI/CD in VSTS. So tests run there. How does configuration works during the build? How do I access app settings? How should I register MyOptions in the tests? Lots of questions...

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: ed39cb0e-aa85-8bba-ffb4-859b56a60723
* Version Independent ID: dc7e3449-650b-d117-5eac-69e5ccb250cb
* Content: [Options pattern in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-2.1#accessing-options-during-startup)
* Content Source: [aspnetcore/fundamentals/configuration/options.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/fundamentals/configuration/options.md)
* Product: **aspnet-core**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by guardrex** [comment]
Hi @alvipeo ... Consult the integration testing topic (and other testing topics): https://docs.microsoft.com/aspnet/core/test/integration-tests

**Comment by guardrex** [comment]
@alvipeo Also note that we don't have the staffing here to field general development questions. We recommend asking for general development support on a support forum, such as Stack Overflow, or a support chat, such as Gitter or Slack.

**Comment by alvipeo** [comment]
That's not general questions. These are related to creating tests and using configuration for the tests in different environments. Those topics should be covered in the docs.

**Comment by guardrex** [comment]
If you're referring to mocking services for integration testing, it's described here ...

https://docs.microsoft.com/aspnet/core/test/integration-tests#inject-mock-services

You'll just mock up your configuration and/or options services and build your tests from there. You'll follow exactly the same pattern as is shown for the `IQuoteService` example.

... or were you more interested in a unit testing approach?

I'd probably just create a ConfigurationBuilder, read the app's sources using the same config providers as the app, and check the values meet predefined criteria. Taking it a step further for options: Bind configuration to an app options class in the test, and then check the options are as you expect.

There might be a more sophisticated approach. However if the idea is just that you want to confirm that whatever config the app is receiving from whatever config sources it's using contains valid configuration (and options binding produces the desired result), that would be a quick and easy way to check that.

We don't recommend testing the configuration/options systems themselves ... they're **heavily tested** by the framework.

We link out for most of our unit testing coverage, so we'd need to discuss further with management if we want to try and adopt additional unit testing guidance here. I'm not opposed to it, but it would probably be a low priority item unless the MS engineers ask us to *make it happen*.

RE: CI/CD/VSTS ... It's not a priority in this doc set, which deals primarily with the ASP.NET Core aspects of testing. Try over in the *Test your code* section of the TFS/VSTS doc set. https://docs.microsoft.com/en-us/vsts/?view=vsts#pivot=start&panel=start-all

For example ... https://docs.microsoft.com/vsts/pipelines/test/getting-started-with-continuous-testing?view=vsts

It should (*in theory* :smile:) detect and run your tests just as if you were doing the testing yourself. If any of the tests fail, the build fails. It should 🤞 ✨ **_Just Work_**:tm: :sparkles:.

I will mark this issue up for a closer look, but it could take some time before we can get to it given the current workload and priorities. Feel free to continue to discuss your progress with me here on the issue.

**Comment by guardrex** [comment]
We're going to pursue additional updates on periodic reviews of the testing topics. If we need to re-open the issue for further discussion, we can do so.
